torchvision/models/maxvit.py

```diff
@@ -0,0 +1,595 @@
+        in_channels: int,
+        out_channels: int,
+        expansion_ratio: float,
+        squeeze_ratio: float,
+        stride: int,
+        activation_fn: Callable[..., nn.Module],
+        normalization_fn: Callable[..., nn.Module],
+    ) -> None:
+        super().__init__()
+
+        proj: Sequence[nn.Module]
+        self.proj: nn.Module
+
+        should_proj = stride != 1 or in_channels != out_channels
+        if should_proj:
+            proj = [nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=1, bias=False)]
+            if stride == 2:
+                proj = [nn.AvgPool2d(kernel_size=3, stride=stride, padding=1)] + proj  # type: ignore
+            self.proj = nn.Sequential(*proj)
+        else:
+            self.proj = nn.Identity()  # type: ignore
+
+        mid_channels = int(out_channels * expansion_ratio)
+        sqz_channels = int(mid_channels * squeeze_ratio)
+
+        _layers = OrderedDict()
+        _layers["pre_norm"] = normalization_fn(in_channels)
+        _layers["conv_a"] = Conv2dNormActivation(
+            in_channels, mid_channels, 1, 1, 0, activation_layer=activation_fn, norm_layer=normalization_fn
+        )
+        _layers["conv_b"] = Conv2dNormActivation(
+            mid_channels,
+            mid_channels,
+            3,
+            stride,
+            1,
+            activation_layer=activation_fn,
+            norm_layer=normalization_fn,
+            groups=mid_channels,
+        )
+        _layers["squeeze_excitation"] = SqueezeExcitation(mid_channels, sqz_channels)
+        _layers["conv_c"] = nn.Conv2d(in_channels=mid_channels, out_channels=out_channels, kernel_size=1, bias=False)
+
+        self.layers = nn.Sequential(_layers)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.layers(x) + self.proj(x)
+
+
+class RelativePositionalMultiHeadAttention(nn.Module):
+    def __init__(
+        self,
+        feat_dim: int,
+        head_dim: int,
+        max_seq_len: int,
+    ) -> None:
+        super().__init__()
+
+        if feat_dim % head_dim != 0:
+            raise ValueError(f"feat_dim: {feat_dim} must be divisible by head_dim: {head_dim}")
+
+        self.n_heads = feat_dim // head_dim
+        self.head_dim = head_dim
+        self.size = int(math.sqrt(max_seq_len))
+        self.max_seq_len = max_seq_len
+
+        self.to_qkv = nn.Linear(feat_dim, self.n_heads * self.head_dim * 3)
+        self.scale_factor = feat_dim**-0.5
+
+        self.merge = nn.Linear(self.head_dim * self.n_heads, feat_dim)
+        self.positional_bias = nn.parameter.Parameter(
+            torch.zeros(((2 * self.size - 1) * (2 * self.size - 1), self.n_heads), dtype=torch.float32),
+        )
+
+        self.register_buffer("relative_position_index", get_relative_position_index(self.size, self.size))
+
+        # initialize with truncated normal the bias
+        self.positional_bias.data.normal_(mean=0, std=0.02)
+
+    def _get_relative_positional_bias(self) -> torch.Tensor:
```

Comment:
Nit: This is a private method used once. Do we really need it?

Reply:
Not necessarily, it can be directly included in the forward pass of the attention op. I've left as is for readability purposes on the user's side so it can be a bit more straightforward on how the relative bias interacts with the dot products.

Reply:
Typically we avoid private methods that are used once but no strong opinions from my side. Up to you.